src/main/java/org/elasticsearch/cloud/aws/blobstore/DefaultS3OutputStream.java

```diff
@@ -120,7 +126,29 @@ protected void doUpload(S3BlobStore blobStore, String bucketName, String blobNam
             md.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
         }
         md.setContentLength(length);
-        blobStore.client().putObject(bucketName, blobName, is, md);
+
+        InputStream inputStream = is;
+
+        // We try to compute a MD5 while reading it
+        MessageDigest messageDigest = null;
+        try {
+            messageDigest = MessageDigest.getInstance("MD5");
+            inputStream = new DigestInputStream(is, messageDigest);
+        } catch (NoSuchAlgorithmException e) {
+            logger.debug("can not check blob upload. MD5 algorithm not supported");
+        }
```

Comment:
In this case I would prefer:

    } catch (NoSuchAlgorithmException impossible) {
        // Every implementation of the Java platform is required to support MD5 (see MessageDigest)
        throw new RuntimeException(impossible);
    }

Then, we don't have to handle the null case afterwards or any of that and it makes the code easier to read.